### ISSUE TITLE
batch processing + extremely slow on new / some samples

### ISSUE THREAD

**Issue by Respaired** [YangXusheng-yxs/CodecFormer_5Hz#1]
awesome work.
but i have a lot of trouble using this model in a batched mode. it's always stuck. after the first step.

i am using pytorch 2.9, python 3.11, arm cpu, GH200 ( i don't see your `requirements.sh`)

is there any pre-defined way to batch extract the discrete codes? 

also, everytime i am trying to extract a new sample, it is extremely slow. minutes of slowness if not stuck all together...
to be more specific : 

```tokenizer.hy_tokenize.encode```
this guy always hangs.

I'd appreciate your help
---
Edit: 
ok i found out torch automatically compiles it for some reason. the fix is just forcing eager:
```
import torch._dynamo
torch._dynamo.config.disable = True
```

**Comment by YangXusheng-yxs** [comment]
Thanks for the positive feedback. I am sorry for the delayed response as I've been busy with other things.

My environment is based on the configuration in uniaudio (https://github.com/yangdongchao/UniAudio/blob/main/UniAudio/requirement.sh), specifically, I used torch 2.3.1+cu118, Python 3.8, on an H20 GPU.

Have you solved this problem? How do you extract discrete codes in batch mode? Could you show me a code snippet or send me your audio file? I can try the inference test.

**Comment by Respaired** [comment]
Likewise Sorry for the late reply.

yes I've fixed the issue. by using : 
```
import torch._dynamo
torch._dynamo.config.disable = True
```
it seems the a section of the code base relies on torch compilation.
I typically extract in batches by padding the input audio and then remove the padded parts right before saving. here's the script i used to extract the codes:

```python
import torch._dynamo
torch._dynamo.config.disable = True

import os
import sys
import torch
import numpy as np
import pandas as pd
import librosa
import warnings
import torch.multiprocessing as mp
from tqdm import tqdm
import logging
import time



USE_CSV = True
DATASET_PATH = "xxx"
CSV_PATH = "/home/ubuntu/tts/new_age/transcribed/final_clean/final_final_final/KHALAS_03_FIL.csv"
BASE_INPUT_DIR = "/home/ubuntu/tts/Arabic_bu"
OUTPUT_DIR = "/home/ubuntu/tts/Arabic_bu/Ucodec_codes"

CONF_PATH = '/home/ubuntu/tts/Navadec_5hz/audio_processing/CodecFormer_5Hz/weights/5Hz_16rvq_c4096_config.json'
CKPT_PATH = '/home/ubuntu/tts/Navadec_5hz/audio_processing/CodecFormer_5Hz/weights/5Hz_16rvq_c4096_generator-650000.pt'

TARGET_SR = 16000
DURATION = 29.0
DOWN_FACTOR = 3200
BATCH_SIZE = 32

NUM_GPUS = 4
NUM_LOADERS = 8
QUEUE_SIZE = 512

def get_output_path(original_path, base_input_dir, output_root):
    original_path = os.path.abspath(original_path)
    base_input_dir = os.path.abspath(base_input_dir)
    
    if original_path.startswith(base_input_dir):
        rel_path = os.path.relpath(original_path, base_input_dir)
    else:
        rel_path = os.path.basename(original_path)
    
    name_only, _ = os.path.splitext(rel_path)
    return os.path.join(output_root, f"{name_only}.pt")


def split_list(lst, n_parts):

    k, m = divmod(len(lst), n_parts)
    return [lst[i * k + min(i, m):(i + 1) * k + min(i + 1, m)] for i in range(n_parts)]


def audio_loader_process(file_list, queue, base_input_dir, output_dir, done_counter):

    for filepath in file_list:
        try:
            out_path = get_output_path(filepath, base_input_dir, output_dir)
            
            if os.path.exists(out_path):
                with done_counter.get_lock():
                    done_counter.value += 1
                continue
            
            if not os.path.exists(filepath):
                with done_counter.get_lock():
                    done_counter.value += 1
                continue

            y, _ = librosa.load(filepath, sr=TARGET_SR, duration=DURATION, mono=True)
            
            if len(y) < 1000:
                with done_counter.get_lock():
                    done_counter.value += 1
                continue
            
            queue.put((y.astype(np.float32), len(y), out_path))
            with done_counter.get_lock():
                done_counter.value += 1
            
        except Exception:
            with done_counter.get_lock():
                done_counter.value += 1



def consumer_worker(rank, gpu_id, input_queue, num_loaders, loaders_done):
    torch.set_num_threads(2)
    warnings.filterwarnings("ignore")
    logging.getLogger("pynvml").setLevel(logging.ERROR)
    
    sys.path.append("/home/ubuntu/tts/Navadec_5hz/audio_processing/CodecFormer_5Hz")
    try:
        from tools.tokenizer.soundstream.AudioTokenizer_HY import HY_Tokenizer
    except ImportError as e:
        print(f"[Worker {rank}] Import Error: {e}")
        return

    device = torch.device(f'cuda:{gpu_id}')
    try:
        tokenizer = HY_Tokenizer(config_path=CONF_PATH, ckpt_path=CKPT_PATH, device=device).to(device)
        tokenizer = torch.compile(tokenizer)
        tokenizer.eval()
        print(f"[Worker {rank}] Initialized on GPU {gpu_id}")
    except Exception as e:
        print(f"[Worker {rank}] Model Init Failed on GPU {gpu_id}: {e}")
        return

    batch_buffer = []
    
    while True:

        if loaders_done.value >= num_loaders and input_queue.empty():
            if batch_buffer:
                process_batch(tokenizer, batch_buffer, device)
            break
        
        try:
            item = input_queue.get(timeout=0.01)
            wav_np, original_len, out_path = item
            
            if os.path.exists(out_path):
                continue
            
            batch_buffer.append({
                'wav': wav_np,
                'len': original_len,
                'path': out_path
            })
            
            if len(batch_buffer) >= BATCH_SIZE:
                process_batch(tokenizer, batch_buffer, device)
                batch_buffer = []
                
        except Exception:
            continue
    
    print(f"[Worker {rank}] Finished on GPU {gpu_id}")


def process_batch(model, batch_data, device):
    if not batch_data:
        return

    tensors = []
    output_paths = []
    original_lens = []
    
    for item in batch_data:
        wav_np = item['wav']
        wav = torch.from_numpy(wav_np).float().to(device)
        
        if wav.ndim == 1:
            wav = wav.unsqueeze(0)
            
        tensors.append(wav.squeeze(0))
        output_paths.append(item['path'])
        original_lens.append(item['len'])

    padded = torch.nn.utils.rnn.pad_sequence(tensors, batch_first=True)
    padded = padded.unsqueeze(1)

    try:
        with torch.no_grad():
            encoded_raw = model.hy_tokenize.encode(padded)
            
        for i, path in enumerate(output_paths):
            item_codes = encoded_raw[:, i, :]
            
            orig_samples = original_lens[i]
            expected_frames = int(np.ceil(orig_samples / DOWN_FACTOR))
            actual_frames = min(expected_frames, item_codes.shape[1])
            
            final_code = item_codes[:, :actual_frames]
            final_cpu = final_code.to(torch.int16).cpu()
            
            os.makedirs(os.path.dirname(path), exist_ok=True)
            torch.save(final_cpu, path)
            
    except Exception as e:
        print(f"Error in batch inference: {e}")

def load_csv_dataset(csv_path):
    print(f"Loading CSV Dataset: {csv_path}")
    
    if not os.path.exists(csv_path):
        raise FileNotFoundError(f"CSV file not found: {csv_path}")
    
    df = pd.read_csv(csv_path)
    
    if 'filename' not in df.columns:
        raise ValueError("CSV must contain a 'filename' column")
    
    paths = df['filename'].dropna().tolist()
    print(f"Found {len(paths)} files in CSV")
    return paths


def loader_wrapper(file_list, queue, base_dir, out_dir, done_counter, loaders_done):
    audio_loader_process(file_list, queue, base_dir, out_dir, done_counter)
    with loaders_done.get_lock():
        loaders_done.value += 1


def main():

    
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    mp.set_start_method('spawn', force=True)
    

    queue = mp.Queue(maxsize=QUEUE_SIZE)
    done_counter = mp.Value('i', 0)
    loaders_done = mp.Value('i', 0)

    file_paths = load_csv_dataset(CSV_PATH)
    total = len(file_paths)
    

    file_splits = split_list(file_paths, NUM_LOADERS)
    
    print(f"\nStarting {NUM_LOADERS} audio loaders and {NUM_GPUS} GPU consumers...")
    start_time = time.time()
    

    consumers = []
    for gpu_id in range(NUM_GPUS):
        p = mp.Process(
            target=consumer_worker,
            args=(gpu_id, gpu_id, queue, NUM_LOADERS, loaders_done)
        )
        p.start()
        consumers.append(p)
    
    loaders = []
    for i in range(NUM_LOADERS):
        p = mp.Process(
            target=loader_wrapper,
            args=(file_splits[i], queue, BASE_INPUT_DIR, OUTPUT_DIR, done_counter, loaders_done)
        )
        p.start()
        loaders.append(p)
    

    pbar = tqdm(total=total, desc="Processing")
    last_count = 0
    
    while any(p.is_alive() for p in consumers):
        current = done_counter.value
        if current > last_count:
            pbar.update(current - last_count)
            last_count = current
        time.sleep(0.1)
    
    pbar.update(total - last_count)
    pbar.close()
    
 
    for p in loaders:
        p.join()
    for p in consumers:
        p.join()
    
    elapsed = time.time() - start_time
    print(f"\n{'=' * 60}")
    print(f"COMPLETED in {elapsed:.1f}s ({elapsed/60:.1f} min)")
    print(f"{'=' * 60}")


if __name__ == "__main__":
    main()
```
